managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -1852,6 +1852,51 @@ public void findEntryFailed(ManagedLedgerException exception, Optional<Position>
         return future;
     }
 
+    @Override
+    public CompletableFuture<Position> asyncReverseFindPositionOneByOne(Predicate<Entry> predicate) {
```

Comment:
Maybe `asyncFindLastValidPosition` or something else is better

Reply:
I want to show two key points in the method name: `Reverse` and `OneByOne`

Reply:
I also support the `asyncFindLastValidPosition` name. `Reverse` and `OneByOne` just describe how do you achieve the goal, but `findLastValidPosition` describes what goal you want to achieve.

If the way of how to achieve the goal changes in future, the method name might not be proper.

Reply:
+1 for `asyncFindLastValidPosition`. We are defining the interface method instead of the implementation. `asyncReverseFindPositionOneByOne` is to reflect the implementation details while `asyncFindLastValidPosition` is more suitable for the interface.

Reply:
We already have `findNewestMatching` method in the ManageCursor API. We can just use the ReadonlyCursor for find the the available position. So that we don't need to make any changes to the ManagedLedger API.

Reply:
It might not be worth creating another cursor.

I think we can use add a static method because all the methods used are public or package visiable.

```java
@InterfaceStability.Evolving
public class ManagedLedgerImplUtils {

    public static CompletableFuture<Position> asyncGetLastValidPosition(final ManagedLedgerImpl ledger,
                                                                        final Predicate<Entry> predicate) {
```

Reply:
> We already have findNewestMatching method in the ManageCursor API. We can just use the ReadonlyCursor for find the the available position. So that we don't need to make any changes to the ManagedLedger API.

@codelipenghui Here need `Reverse` find, and this method doesn't seem match.

@BewareMyPower @RobertIndie @tjiuming What do you think call `asyncReverseFindLastValidPosition`?  I think `Reverse` should be part of the method name.


> I think we can use add a static method because all the methods used are public or package visible.

Is there any harm to adding an interface? I'd prefer to embody it on the interface, which shows the various use cases of Pulsar for the `ManangerLedger`.

Reply:
> Is there any harm to adding an interface?

See https://github.com/apache/pulsar/pull/22572#discussion_r1582003060

```java
  /**
   * Can evolve while retaining compatibility for minor release boundaries.;
   * can break compatibility only at major release (ie. at m.0).
   */
```

So you have to use a default method or a static method.

> What do you think call asyncReverseFindLastValidPosition? I think Reverse should be part of the method name.

`Reverse` is still redundant. You can take the Java standard API for example:

```java
        String s = "hello";
        s.lastIndexOf('l');
```

It's `lastIndexOf`, no something like `reverse`. Because `Last` is already a hint for reverse finding. Python string API has `index` and `rindex`, where `r` is the hint for `reverse`. But it does not have any other hint like `last`.

Reply:
Regarding implementation, actually we can use sequential find on the active ledger rather than the reverse finding because we already have the cache. We only need reverse finding when all entries in the active ledger are not available.

So "Reverse" is not required.

Reply:
Thanks for explaining, I will accept this suggestion.

https://github.com/apache/pulsar/pull/22572#discussion_r1582006268

And, I will keep `reverse ` find, Because in most cases, the last entity can meet the predicate.

Reply:
> @codelipenghui Here need Reverse find, and this method doesn't seem match.

Oh I see. Thanks for the explanation. A new method is good to me.

And I think we can follow @BewareMyPower's suggestion to add a Until class. 

> Regarding implementation, actually we can use sequential find on the active ledger rather than the reverse finding because we already have the cache. We only need reverse finding when all entries in the active ledger are not available.

I didn't get the point. Even if the entries are cached, reverse reading is also more efficient for getting the last valid entry?

Reply:
It's just an example to show the implementation could vary. Not a suggestion to the actual implementation. I meant
1. `read(activeLedger)` returns a list of entries (`entries`)
2. `for (int i = entries.size() - 1; i >= 0; i--)` reverse find

not

```
while (true) {
  var position = getPreviousPosition(position);
  var entry = readOneEntry(position);
  /* ... */
}
```

The code above is pseudo code. When I mentioned the implementation, I didn't look into the details of the cache. So please don't take serious on my example.